operator/src/lib.rs

```diff
@@ -282,8 +251,163 @@ impl SparkState {
+    /// That means not "terminating" and "ready"
+    /// We always make sure that all masters are running before any workers / history-servers
+    async fn check_pods_up_and_running(&self) -> SparkReconcileResult {
+        trace!("Reconciliation: Checking if all pods are up and running");
+
+        if let Some(pod_info) = &self.pod_information {
+            // master nodes have priority
+            if !self.pods_up_and_running(pod_info.get_all_pods(Some(SparkNodeType::Master))) {
+                return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+            }
+
+            // worker / history-server are treated equally
+            if !self.pods_up_and_running(pod_info.get_all_pods(Some(SparkNodeType::Worker)))
+                || !self
+                    .pods_up_and_running(pod_info.get_all_pods(Some(SparkNodeType::HistoryServer)))
+            {
+                return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+            }
+        }
+
+        Ok(ReconcileFunctionAction::Continue)
+    }
+    /// Returns true only if any pod status is neither "terminating" or anything other than "ready"
+    fn pods_up_and_running(&self, pods: Vec<Pod>) -> bool {
+        let mut terminated_pods = vec![];
+        let mut started_pods = vec![];
+        for pod in &pods {
+            if finalizer::has_deletion_stamp(pod) {
+                terminated_pods.push(Meta::name(pod));
+            }
+
+            if !podutils::is_pod_running_and_ready(pod) {
+                started_pods.push(Meta::name(pod));
+            }
+        }
+
+        // If pods are currently terminating, wait until they are done terminating.
+        // (this could be for restarts or upgrades)
+        if !terminated_pods.is_empty() {
+            info!("Waiting for Pod(s) to be terminated: {:?}", terminated_pods);
+            return false;
+        }
+
+        // At the moment we'll wait for all pods of a certain type (especially master),
+        // to be available and ready before we might enact any changes to existing ones.
+        if !started_pods.is_empty() {
+            info!(
+                "Waiting for Pod(s) to be running and ready: {:?}",
+                started_pods
+            );
+            return false;
+        }
+
+        true
+    }
+
+    /// Process the cluster status version for upgrades/downgrades
+    /// When creating the status for the first time, current_version equals target_version
+    /// If we need to upgrade/downgrade the version we set the target version to the required upgrade/downgrade
+    /// Once all pods are running with the upgraded/downgraded version we will set current_version to the new version
+    pub async fn process_version_update(&mut self) -> SparkReconcileResult {
+        if let (Some(status), Some(pod_info)) = (&self.status, &self.pod_information) {
+            // We have 3 cases to cover
```

Comment:
See my comments in the CRD about this, I'm not going to comment further on this here because I think we should discuss the general principle first.